Window Controls

CubesterYT opened the pull request:
Resolves #360
Resolves #454
Resolves #590

Hello all! This is a PR that is in collaboration between me and BlueDome. Me and him have merged ours extensions together and made a even better one! This extension gives you control over everything we could think of. More blocks will be coming soon!

This is the current roster:
![image](https://github.com/TurboWarp/extensions/assets/78769806/a9f610f3-4fdd-45b3-a167-d4bb595938c8)
![image](https://github.com/TurboWarp/extensions/assets/78769806/c6d18b62-ffa4-460f-a1c1-3d9818f6ea04)

LilyMakesThings commented:
COLLAB EXTENSION REALLL

LilyMakesThings commented:
We should really be specifying this does not work in browser and only in electron applications. I can already imagine the issue statements.

CubesterYT commented:
> We should really be specifying this does not work in browser and only in electron applications. I can already imagine the issue statements.

@LilyMakesThings I will be making a documentation page specifying this. Also, after vigorous testing, some blocks do still work in browsers/html. Just the ones that don't affect the page state.

GarboMuffin commented:
put that ones that work in Electron only under a BlockType.LABEL, perhaps

CubesterYT commented:
> put that ones that work in Electron only under a BlockType.LABEL, perhaps

I could, but basically all of them work in electron only. So it could mess up the entire order. So I'll just define in documentation that most work in electron only.

LilyMakesThings commented:
> > put that ones that work in Electron only under a BlockType.LABEL, perhaps
> 
> @GarboMuffin I could, but basically all of them work in electron only. So it could mess up the entire order. So I'll just define in documentation that most work in electron only.

You need to make a label saying the blocks only work in electron. The kind of people who would be making issues about it not working in browser and the kind of people who don't read the documentation.

CubesterYT commented:
> > > put that ones that work in Electron only under a BlockType.LABEL, perhaps
> > 
> > @GarboMuffin I could, but basically all of them work in electron only. So it could mess up the entire order. So I'll just define in documentation that most work in electron only.
> 
> You need to make a label saying the blocks only work in electron. The kind of people who would be making issues about it not working in browser are the kind of people who don't read the documentation.

Ahhhh like that one other extension I forgot the name of that had a label stating something important. Alright I will do that tomorrow with some more block additions and documentation and anything else. Then everything should be ready.

CubesterYT commented:
> Could this then help with #360?

@UndermanReal Yes actually. I haven't added it yet because I'm outside. But I will be adding a block called "match stage size" that will change the window aspect ratio to match the canvas.

CubesterYT commented:
@GarboMuffin @LilyMakesThings Wanted to update you guys on something. After further testing, Window Controls works perfectly in Electron AND Pop Up HTML. Only certain ones work in browser/browser HTML.

CubesterYT commented:
@GarboMuffin, Window Controls is ready for review/ready to be merged.

echoless3484 commented:
I feel like these blocks need a spot, while deprecating the enter and exit fullscreen blocks
![scratchblocks (1)](https://github.com/TurboWarp/extensions/assets/109392419/62de8c1b-ea3a-4b00-b261-36f15da83a48)

CubesterYT commented:
> I feel like these blocks need a spot, while deprecating the enter and exit fullscreen blocks
> ![scratchblocks (1)](https://github.com/TurboWarp/extensions/assets/109392419/62de8c1b-ea3a-4b00-b261-36f15da83a48)
> 

@echoless3484, let me explain some things. The set Fullscreen to true/false block doesn't make sense on how the Fullscreen API works. Fullscreen API is a tricky thing to explain, as the API uses different sets to make it function. Fullscreen API functions using the sub window API called "window.document" which has its own set of APIs. To enter Fullscreen, it uses ANOTHER SUB API within "window.document", which is "window.document.documentElement", and to enter Fullscreen, we use "document.documentElement.requestFullscreen()". This is because ".documentElement.requestFullscreen" can be added to any element within the window, not just the window. Now for exiting, that doesn't even use "document.documentElement", but uses just "window.document" for some odd reason. So to exit we run "document.exitFullscreen()". So we don't just set Fullscreen to true or false. So enter/exit cannot be deprecated, nor can "set Fullscreen to true/false" be added. BUT... Based on the reporter named simply "Fullscreen" that most likely returns if the window is in Fullscreen or not IS POSSIBLE. There is an official API that returns true or false in correlation to the window being Fullscreen or not. And I actually did have plans to add a boolean named, "is window fullscreen?", but I had forgotten to add it. But I will be adding this in the next version of Window Controls.

You can find more details about the Fullscreen API in depth here: https://developer.mozilla.org/en-US/docs/Web/API/Fullscreen_API

creepersaur commented:
Oh wow, amazing. Is there a link or github page for the newest extension file?

echoless3484 commented:
> Oh wow, amazing. Is there a link or github page for the newest extension file?

Go to the top and click on "Window Controls" in the "Window Controls > Master" boxes

creepersaur commented:
Now what's left is to be able to fullscreen, get the width and height of the display/screen, and to minimize, maximize, and change the ability of a window to be resizeable.

Also imagine what we could do if we could open apps using scratch, like export 2 exe projects, and run the exe/open the project using scratch code.

echoless3484 commented:
> Now what's left is to be able to fullscreen, get the width and height of the display/screen, and to minimize, maximize, and change the ability of a window to be resizeable.
> 
> Also imagine what we could do if we could open apps using scratch, like export 2 exe projects, and run the exe/open the project using scratch code.

this extension already has fullscreen supports

CubesterYT commented:
> Now what's left is to be able to fullscreen, get the width and height of the display/screen, and to minimize, maximize, and change the ability of a window to be resizeable.
> 
> Also imagine what we could do if we could open apps using scratch, like export 2 exe projects, and run the exe/open the project using scratch code.

@creepersaur This extension can already enter and exit fullscreen. It can also already get the width and height of the screen, hence the screen x/y reporters. I'm not exactly sure what you mean by setting the ability of a window to be resizeable. And finally, I already have plans to add blocks that can minimize and maximize the window, but they won't be added among a ton others once I fully have a list of which BrowserWindow apis I want GarboMuffin to add.

creepersaur commented:
> > Now what's left is to be able to fullscreen, get the width and height of the display/screen, and to minimize, maximize, and change the ability of a window to be resizeable.
> > Also imagine what we could do if we could open apps using scratch, like export 2 exe projects, and run the exe/open the project using scratch code.
> 
> @creepersaur This extension can already enter and exit fullscreen. It can also already get the width and height of the screen, hence the screen x/y reporters. I'm not exactly sure what you mean by setting the ability of a window to be resizeable. And finally, I already have plans to add blocks that can minimize and maximize the window, but they won't be added among a ton others once I fully have a list of which BrowserWindow apis I want GarboMuffin to add.

Ohh. I wasn't using the latest extension file. Now I see the screen w,h and the fullscreen, etc. This is very cool. The window resizeability means that currently, the user can drag on the corners of the application to change the size (w,h) of the window. Which you may not want in all cases. So the ability to stop the user from doing that could help. And also I don't know if this is possible, but removing the window border (including the edges of the application window and the top bar) so you could just have a plain window application, and then add your own border and things if you want.

CubesterYT commented:
> > > Now what's left is to be able to fullscreen, get the width and height of the display/screen, and to minimize, maximize, and change the ability of a window to be resizeable.
> > > Also imagine what we could do if we could open apps using scratch, like export 2 exe projects, and run the exe/open the project using scratch code.
> > 
> > 
> > @creepersaur This extension can already enter and exit fullscreen. It can also already get the width and height of the screen, hence the screen x/y reporters. I'm not exactly sure what you mean by setting the ability of a window to be resizeable. And finally, I already have plans to add blocks that can minimize and maximize the window, but they won't be added among a ton others once I fully have a list of which BrowserWindow apis I want GarboMuffin to add.
> 
> Ohh. I wasn't using the latest extension file. Now I see the screen w,h and the fullscreen, etc. This is very cool. The window resizeability means that currently, the user can drag on the corners of the application to change the size (w,h) of the window. Which you may not want in all cases. So the ability to stop the user from doing that could help. And also I don't know if this is possible, but removing the window border (including the edges of the application window and the top bar) so you could just have a plain window application, and then add your own border and things if you want.

@creepersaur Ahhh I see. I will try to add a block that counters window size change, but you can currently use a method where if the program detects the window size changes, then you can set the size back to the dimensions you want. And about window border, I'm not sure if it's possible, I thought of it, but from what I'm seeing, it may be only for .appx apps.

BlueDome77 commented:
> > > > Now what's left is to be able to fullscreen, get the width and height of the display/screen, and to minimize, maximize, and change the ability of a window to be resizeable.
> > > > Also imagine what we could do if we could open apps using scratch, like export 2 exe projects, and run the exe/open the project using scratch code.
> > > 
> > > 
> > > @creepersaur This extension can already enter and exit fullscreen. It can also already get the width and height of the screen, hence the screen x/y reporters. I'm not exactly sure what you mean by setting the ability of a window to be resizeable. And finally, I already have plans to add blocks that can minimize and maximize the window, but they won't be added among a ton others once I fully have a list of which BrowserWindow apis I want GarboMuffin to add.
> > 
> > 
> > Ohh. I wasn't using the latest extension file. Now I see the screen w,h and the fullscreen, etc. This is very cool. The window resizeability means that currently, the user can drag on the corners of the application to change the size (w,h) of the window. Which you may not want in all cases. So the ability to stop the user from doing that could help. And also I don't know if this is possible, but removing the window border (including the edges of the application window and the top bar) so you could just have a plain window application, and then add your own border and things if you want.
> 
> @creepersaur Ahhh I see. I will try to add a block that counters window size change, but you can currently use a method where if the program detects the window size changes, then you can set the size back to the dimensions you want. And about window border, I'm not sure if it's possible, I thought of it, but from what I'm seeing, it may be only for .appx apps.

Maybe [this ](https://www.electronjs.org/docs/latest/tutorial/window-customization)could help?

CubesterYT commented:
> > > > > Now what's left is to be able to fullscreen, get the width and height of the display/screen, and to minimize, maximize, and change the ability of a window to be resizeable.
> > > > > Also imagine what we could do if we could open apps using scratch, like export 2 exe projects, and run the exe/open the project using scratch code.
> > > > 
> > > > 
> > > > @creepersaur This extension can already enter and exit fullscreen. It can also already get the width and height of the screen, hence the screen x/y reporters. I'm not exactly sure what you mean by setting the ability of a window to be resizeable. And finally, I already have plans to add blocks that can minimize and maximize the window, but they won't be added among a ton others once I fully have a list of which BrowserWindow apis I want GarboMuffin to add.
> > > 
> > > 
> > > Ohh. I wasn't using the latest extension file. Now I see the screen w,h and the fullscreen, etc. This is very cool. The window resizeability means that currently, the user can drag on the corners of the application to change the size (w,h) of the window. Which you may not want in all cases. So the ability to stop the user from doing that could help. And also I don't know if this is possible, but removing the window border (including the edges of the application window and the top bar) so you could just have a plain window application, and then add your own border and things if you want.
> > 
> > 
> > @creepersaur Ahhh I see. I will try to add a block that counters window size change, but you can currently use a method where if the program detects the window size changes, then you can set the size back to the dimensions you want. And about window border, I'm not sure if it's possible, I thought of it, but from what I'm seeing, it may be only for .appx apps.
> 
> Maybe [this ](https://www.electronjs.org/docs/latest/tutorial/window-customization)could help?

@BlueDome77 I already know of this, and I have it planned to add, but again, like I said a few comments ago, I'm still talking to GarboMuffin of which BrowserWindow APIs I want him to add. Once he does, I can add tons of new blocks, like min/max the window, and tons more.

BlueDome77 commented:
> 

My bad for not reading the previous comments

CubesterYT commented:
> > 
> 
> My bad for not reading the previous comments

It's fine

CubesterYT commented:
Ok, NOW Window Controls is fully ready for review.

NexusKitten commented:
Why'd you shoot down #778 and then add a block from it into your extension without any credit? I feel like it would be fine as it's own extension to be honest, it feels too different to this one.

BlueDome77 commented:
> Why'd you shoot down #778 and then add a block from it into your extension without any credit? I feel like it would be fine as it's own extension to be honest, it feels too different to this one.

It isn't necessary to have an extension that has a couple of blocks that are already implemented better in another extension.

NexusKitten commented:
> > Why'd you shoot down #778 and then add a block from it into your extension without any credit? I feel like it would be fine as it's own extension to be honest, it feels too different to this one.
> 
> It isn't necessary to have an extension that has a couple of blocks that are already implemented better in another extension.

They weren't implemented better though, given that OP had to add a crucial block from the other one. It was implemented, just not implemented well. That's not to say the other one is infinitely better, however telling someone to remove their PR because they already had those features, only to add those features later, is rather rude.

CubesterYT commented:
> Why'd you shoot down #778 and then add a block from it into your extension without any credit? I feel like it would be fine as it's own extension to be honest, it feels too different to this one. 

@NexusKitten, I had already had the boolean that checks for fullscreen status for a while now, I never added it in case this PR was going to be merged. Since it hasn't yet, I added the new boolean, and I did NOT take anything from the other PR, if you read the above comments about me explaining about fullscreen api, I had mentioned I had a an api ready to check fullscreen. Do not make such claims of me taking ones code.

CubesterYT commented:
> > > Why'd you shoot down #778 and then add a block from it into your extension without any credit? I feel like it would be fine as it's own extension to be honest, it feels too different to this one.
> > 
> > 
> > It isn't necessary to have an extension that has a couple of blocks that are already implemented better in another extension.
> 
> They weren't implemented better though, given that OP had to add a crucial block from the other one. It was implemented, just not implemented well. That's not to say the other one is infinitely better, however telling someone to remove their PR because they already had those features, only to add those features later, is rather rude.

@NexusKitten, another thing, the blocks are made better by adding failsafe methods that don't give off errors, and allow for the blocks to run better.

NexusKitten commented:
> > Why'd you shoot down #778 and then add a block from it into your extension without any credit? I feel like it would be fine as it's own extension to be honest, it feels too different to this one.
> 
> @NexusKitten, I had already had the boolean that checks for fullscreen status for a while now, I never added it in case this PR was going to be merged. Since it hasn't yet, I added the new boolean, and I did NOT take anything from the other PR, if you read the above comments about me explaining about fullscreen api, I had mentioned I had a an api ready to check fullscreen. I've always been prepared to add the new block. Do not make such claims of me taking ones code.

I never once claimed you stole code, all I'm saying is that it's a little rude to shoot down someone else's PR (that already had a thumbnail and all, other than bug fixing, is completely ready to merge) just to add blocks from that PR into your extension without even a comment. It still feels like it either belongs in it's own extension, or (with credit to the extension creator and you) Browser Utilities, as fullscreen has nothing to do with "Window Controls."

CubesterYT commented:
> > > Why'd you shoot down #778 and then add a block from it into your extension without any credit? I feel like it would be fine as it's own extension to be honest, it feels too different to this one.
> > 
> > 
> > @NexusKitten, I had already had the boolean that checks for fullscreen status for a while now, I never added it in case this PR was going to be merged. Since it hasn't yet, I added the new boolean, and I did NOT take anything from the other PR, if you read the above comments about me explaining about fullscreen api, I had mentioned I had a an api ready to check fullscreen. I've always been prepared to add the new block. Do not make such claims of me taking ones code.
> 
> I never once claimed you stole code, all I'm saying is that it's a little rude to shoot down someone else's PR (that already had a thumbnail and all, other than bug fixing, is completely ready to merge) just to add blocks from that PR into your extension without even a comment. It still feels like it either belongs in it's own extension, or (with credit to the extension creator and you) Browser Utilities, as fullscreen has nothing to do with "Window Controls."

@NexusKitten, didn't I just say the blocks we're already present in my extension, and have no relation to the other extension at all. I added fullscreen because, like the other blocks, modifies the state of the Window. Fullscreen blocks were one of the first blocks added to this extension, you may even check the commit history.

CubesterYT commented:
Again, recheck everything in this PR before making assumptions with what happened.

NexusKitten commented:
> Again, recheck everything in this PR before making assumptions with what happened.

I'm not making any assumptions. I know you mentioned making a boolean a while ago. Still feels wrong that you shut down the other pull request even though it fits better there. They can fix the bug that shows an error with like 4 lines of code, and I feel like the blocks would belong better there or browser utilities. I mean, it's browser fullscreen; most features in the extension work mostly for electron.

CubesterYT commented:
> > Again, recheck everything in this PR before making assumptions with what happened.
> 
> I'm not making any assumptions. I know you mentioned making a boolean a while ago. Still feels wrong that you shut down the other pull request even though it fits better there. They can fix the bug that shows an error with like 4 lines of code, and I feel like the blocks would belong better there or browser utilities. I mean, it's browser fullscreen; most features in the extension work mostly for electron.

@NexusKitten, if you've been attentive in the Discord Server, or looked at the code for the HTML Documentation, you'll see that every single one of these blocks work outside of Electron, as this extension uses HTML based window api, meaning it works EVERYWHERE. Most blocks don't work in normal browser tabs due to browser security. But after vigorous testing, Window Controls works in Electron, HTML Pop Up tabs, Web Apps, and more.

David-Orangemoon commented:
> > > > put that ones that work in Electron only under a BlockType.LABEL, perhaps
> > > 
> > > 
> > > @GarboMuffin I could, but basically all of them work in electron only. So it could mess up the entire order. So I'll just define in documentation that most work in electron only.
> > 
> > 
> > You need to make a label saying the blocks only work in electron. The kind of people who would be making issues about it not working in browser are the kind of people who don't read the documentation.
> 
> @LilyMakesThings Ahhhh like that one other extension I forgot the name of that had a label stating something important. Alright I will do that tomorrow with some more block additions and documentation and anything else. Then everything should be ready.

Where is that label?

NexusKitten commented:
> > put that ones that work in Electron only under a BlockType.LABEL, perhaps
> 
> @GarboMuffin I could, but basically all of them work in electron only. So it could mess up the entire order. So I'll just define in documentation that most work in electron only.

this you

David-Orangemoon commented:
Also you can't close the window while its moving and the move window to the [position] block seems to do nothing

CubesterYT commented:
> > > put that ones that work in Electron only under a BlockType.LABEL, perhaps
> > 
> > 
> > @GarboMuffin I could, but basically all of them work in electron only. So it could mess up the entire order. So I'll just define in documentation that most work in electron only.
> 
> this you

This is prior to testing vigorously, if you want, I can send a video right now of everything working.

CubesterYT commented:
> > > > > put that ones that work in Electron only under a BlockType.LABEL, perhaps
> > > > 
> > > > 
> > > > @GarboMuffin I could, but basically all of them work in electron only. So it could mess up the entire order. So I'll just define in documentation that most work in electron only.
> > > 
> > > 
> > > You need to make a label saying the blocks only work in electron. The kind of people who would be making issues about it not working in browser are the kind of people who don't read the documentation.
> > 
> > 
> > @LilyMakesThings Ahhhh like that one other extension I forgot the name of that had a label stating something important. Alright I will do that tomorrow with some more block additions and documentation and anything else. Then everything should be ready.
> 
> Where is that label?

It's no longer needed.

David-Orangemoon commented:
> > > > > > put that ones that work in Electron only under a BlockType.LABEL, perhaps
> > > > > 
> > > > > 
> > > > > @GarboMuffin I could, but basically all of them work in electron only. So it could mess up the entire order. So I'll just define in documentation that most work in electron only.
> > > > 
> > > > 
> > > > You need to make a label saying the blocks only work in electron. The kind of people who would be making issues about it not working in browser are the kind of people who don't read the documentation.
> > > 
> > > 
> > > @LilyMakesThings Ahhhh like that one other extension I forgot the name of that had a label stating something important. Alright I will do that tomorrow with some more block additions and documentation and anything else. Then everything should be ready.
> > 
> > 
> > Where is that label?
> 
> It's no longer needed.

why?

NexusKitten commented:
> > > > > > put that ones that work in Electron only under a BlockType.LABEL, perhaps
> > > > > 
> > > > > 
> > > > > @GarboMuffin I could, but basically all of them work in electron only. So it could mess up the entire order. So I'll just define in documentation that most work in electron only.
> > > > 
> > > > 
> > > > You need to make a label saying the blocks only work in electron. The kind of people who would be making issues about it not working in browser are the kind of people who don't read the documentation.
> > > 
> > > 
> > > @LilyMakesThings Ahhhh like that one other extension I forgot the name of that had a label stating something important. Alright I will do that tomorrow with some more block additions and documentation and anything else. Then everything should be ready.
> > 
> > 
> > Where is that label?
> 
> It's no longer needed.

If I click a block and it does nothing, I'd put up an issue on the github. We don't want that to happen.

CubesterYT commented:
> > > > > > > put that ones that work in Electron only under a BlockType.LABEL, perhaps
> > > > > > 
> > > > > > 
> > > > > > @GarboMuffin I could, but basically all of them work in electron only. So it could mess up the entire order. So I'll just define in documentation that most work in electron only.
> > > > > 
> > > > > 
> > > > > You need to make a label saying the blocks only work in electron. The kind of people who would be making issues about it not working in browser are the kind of people who don't read the documentation.
> > > > 
> > > > 
> > > > @LilyMakesThings Ahhhh like that one other extension I forgot the name of that had a label stating something important. Alright I will do that tomorrow with some more block additions and documentation and anything else. Then everything should be ready.
> > > 
> > > 
> > > Where is that label?
> > 
> > 
> > It's no longer needed.
> 
> why?

Because I added documentation specifying the support, and it was supposed to specify it's electron only, but after testing, no need to add it since it works anywhere.

CubesterYT commented:
> Also you can't close the window while its moving and the move window to the [position] block seems to do nothing

What are you using the preset block on?

David-Orangemoon commented:
> > > > > > > > put that ones that work in Electron only under a BlockType.LABEL, perhaps
> > > > > > > 
> > > > > > > 
> > > > > > > @GarboMuffin I could, but basically all of them work in electron only. So it could mess up the entire order. So I'll just define in documentation that most work in electron only.
> > > > > > 
> > > > > > 
> > > > > > You need to make a label saying the blocks only work in electron. The kind of people who would be making issues about it not working in browser are the kind of people who don't read the documentation.
> > > > > 
> > > > > 
> > > > > @LilyMakesThings Ahhhh like that one other extension I forgot the name of that had a label stating something important. Alright I will do that tomorrow with some more block additions and documentation and anything else. Then everything should be ready.
> > > > 
> > > > 
> > > > Where is that label?
> > > 
> > > 
> > > It's no longer needed.
> > 
> > 
> > why?
> 
> Because I added documentation specifying the support, and it was supposed to specify it's electron only, but after testing, no need to add it since it works anywhere.

You should probably fix some of the issues with the blocks
the move window block bricks turbowarp desktop!

Also the move to preset block doesn't seem to work.

CubesterYT commented:
> > > > > > > > > put that ones that work in Electron only under a BlockType.LABEL, perhaps
> > > > > > > > 
> > > > > > > > 
> > > > > > > > @GarboMuffin I could, but basically all of them work in electron only. So it could mess up the entire order. So I'll just define in documentation that most work in electron only.
> > > > > > > 
> > > > > > > 
> > > > > > > You need to make a label saying the blocks only work in electron. The kind of people who would be making issues about it not working in browser are the kind of people who don't read the documentation.
> > > > > > 
> > > > > > 
> > > > > > @LilyMakesThings Ahhhh like that one other extension I forgot the name of that had a label stating something important. Alright I will do that tomorrow with some more block additions and documentation and anything else. Then everything should be ready.
> > > > > 
> > > > > 
> > > > > Where is that label?
> > > > 
> > > > 
> > > > It's no longer needed.
> > > 
> > > 
> > > why?
> > 
> > 
> > Because I added documentation specifying the support, and it was supposed to specify it's electron only, but after testing, no need to add it since it works anywhere.
> 
> You should probably fix some of the issues with the blocks the move window block bricks turbowarp desktop!
> 
> Also the move to preset block doesn't seem to work.

1. what issues? I've tested everything, in all platforms, everything is perfect.
2. If your desktop app is bricked, it's because you made a script using Window Controls that makes the app have high ram usage, which slows the app and computer down, nothing you can do about that.
3. the preset block works, tell me how you're using it.

David-Orangemoon commented:
![block_7_27_2023-11_01_20 PM](https://github.com/TurboWarp/extensions/assets/76855369/ecabe7c7-807c-4ce3-9bb1-ff0be79230d3)
Here is the script I don't know how this would use all 64Gigs of ram

David-Orangemoon commented:
> 

How extensive of testing have you done?

CubesterYT commented:
It's because the blocks work instantly, since you're using it in a forever loop, it moves the window at a speed you can't even comprehend, which can make the window unusable, in order to counter it, at the end of the loop, like under change window, add "wait 0 seconds".

CubesterYT commented:
> > 
> 
> How extensive of testing have you done?

Hours upon hours of testing, so I know my extension inside out, and tried everything.

CubesterYT commented:
> It's because the blocks work instantly, since you're using it in a forever loop, it moves the window at a speed you can't even comprehend, which can make the window unusable, in order to counter it, at the end of the loop, like under change window, add "wait 0 seconds".

I prefer using wait 0.01 seconds for a better effect, but 0 seconds is the base recommendation.

CubesterYT commented:
Let's move this conversation to Discord.

David-Orangemoon commented:
> It's because the blocks work instantly, since you're using it in a forever loop, it moves the window at a speed you can't even comprehend, which can make the window unusable, in order to counter it, at the end of the loop, like under change window, add "wait 0 seconds".

Issue. The forever loop when screen is refreshed waits a frame.

CubesterYT commented:
> > It's because the blocks work instantly, since you're using it in a forever loop, it moves the window at a speed you can't even comprehend, which can make the window unusable, in order to counter it, at the end of the loop, like under change window, add "wait 0 seconds".
> 
> Issue. The forever loop when screen is refreshed waits a frame.

This is something I can't fix, this is just how the API works.

NexusKitten commented:
> > > It's because the blocks work instantly, since you're using it in a forever loop, it moves the window at a speed you can't even comprehend, which can make the window unusable, in order to counter it, at the end of the loop, like under change window, add "wait 0 seconds".
> > 
> > 
> > Issue. The forever loop when screen is refreshed waits a frame.
> 
> This is something I can't fix, this is just how the API works.

Nonetheless, it feels like it's an issue that needs to be addressed. You could:
- Remove the "change by" blocks to discourage the blocks being used in a loop.
- Automatically limit the rate the API can fire in the code
- Ask the user's permission before moving the screen, to avoid random bricking of windows.
If it were up to me, I'd like all 3 to be in place, but ultimately, anything working against bricking the user's window with 6 blocks of code is a step in the right direction.

CST1229 commented:
> > > It's because the blocks work instantly, since you're using it in a forever loop, it moves the window at a speed you can't even comprehend, which can make the window unusable, in order to counter it, at the end of the loop, like under change window, add "wait 0 seconds".
> > 
> > 
> > Issue. The forever loop when screen is refreshed waits a frame.
> 
> This is something I can't fix, this is just how the API works.

You could make the blocks request a screen redraw.

CubesterYT commented:
> > > > It's because the blocks work instantly, since you're using it in a forever loop, it moves the window at a speed you can't even comprehend, which can make the window unusable, in order to counter it, at the end of the loop, like under change window, add "wait 0 seconds".
> > > 
> > > 
> > > Issue. The forever loop when screen is refreshed waits a frame.
> > 
> > 
> > This is something I can't fix, this is just how the API works.
> 
> You could make the blocks request a screen redraw.

Can you tell me how? I will experiment it.

CST1229 commented:
> can you tell me how? I will experiment it.

`Scratch.vm.runtime.requestRedraw()`

CubesterYT commented:
Ok, now Window Controls is TRULY done. It is ready to be merged.

CubesterYT commented:
I give up having to comstantly remove .vscode, which is force adding itself to my repo for some reason.

CubesterYT commented:
Ok, Window Controls is ready to be merged, for real this time.

NexusKitten commented:
Regarding https://github.com/TurboWarp/extensions/pull/778#issuecomment-1663121242

> What do you mean by "unnecessary"? All blocks in Window Controls are necessary, without any of the ones in the current palette, there's no reason for Window Controls to exist.

When @LilyMakesThings said your blocks were unnecessary, what she meant is that the beginning of the block pallet is bloated. Due to lack of dropdowns, the first two sections have 8 and 11 blocks each, when realistically it could function with 3 each. While blocks such as "change window _ by ()" are useful, having them as their own separate blocks bloat the pallet quite a bit, and makes it so you have to scroll quite a bit to get to the other functions of the extension.
> It would be perfect to say that Window Controls will soon be an extension with the most blocks in the entirety of all the extensions available when I'm done with it.

No offense in saying this, but this phrasing makes the extension feel like a Utilities extension. I and most others had assumed the extension's main purpose was to manipulate window position and size, however there's now several other blocks (some of which are already in other extensions) that do not pertain to that theme. I saw where you said that the theme that ties your extension together is window.document, however I believe this theme to be far too vague. It's like justifying an extension because it has blocks, it's technically true, but is far too vague to actually gauge the extension's contents by looking at it.

NexusKitten commented:
All in all I feel as if the best course of action would be to remove the "window title," "window focused," and "window fullscreen" blocks from the extension. All of these blocks are already in other PRs (#614 and #778 respectively), and feel off-topic to the extension's primary function. With it's new patches, the window manipulation blocks work really well, however I feel like the extension should stick to that rather than expand to all things window related.

LilyMakesThings commented:
Following from what @NexusKitten said, the best TurboWarp extensions have the smallest palettes. It's a terrible mindset to aim for quantity over quality, as is very much implied in "Window Controls will soon be an extension with the most blocks of all extensions available".

This palette can produce the exact same functionality of the current extension, but isn't unnecessarily large. Consider limiting the palette.
![image](https://github.com/TurboWarp/extensions/assets/127533508/044d715d-76a1-443b-a578-ff1879e443c2)

CubesterYT commented:
> Regarding [#778 (comment)](https://github.com/TurboWarp/extensions/pull/778#issuecomment-1663121242)
> 
> > What do you mean by "unnecessary"? All blocks in Window Controls are necessary, without any of the ones in the current palette, there's no reason for Window Controls to exist.
> 
> When @LilyMakesThings said your blocks were unnecessary, what she meant is that the beginning of the block pallet is bloated. Due to lack of dropdowns, the first two sections have 8 and 11 blocks each, when realistically it could function with 3 each. While blocks such as "change window _ by ()" are useful, having them as their own separate blocks bloat the pallet quite a bit, and makes it so you have to scroll quite a bit to get to the other functions of the extension.
> 
> > It would be perfect to say that Window Controls will soon be an extension with the most blocks in the entirety of all the extensions available when I'm done with it.
> 
> No offense in saying this, but this phrasing makes the extension feel like a Utilities extension. I and most others had assumed the extension's main purpose was to manipulate window position and size, however there's now several other blocks (some of which are already in other extensions) that do not pertain to that theme. I saw where you said that the theme that ties your extension together is window.document, however I believe this theme to be far too vague. It's like justifying an extension because it has blocks, it's technically true, but is far too vague to actually gauge the extension's contents by looking at it.

@NexusKitten, I don't have much time to explain everything, as it's late night for me, but I specifically state this extension gives you control over every aspect of the Window. This wasn't an extension to just move and resize the window only. Everything you can do to a window is what I want to have in Window Controls. It's my choice because it just makes sense, and I understand about me being able to compress it all, but it still just feels natural not compressing it. Window Controls end goal is to be the go to extension for anything related to the window, as it wants to give you full ***control*** over the window in ***any way***. This is my choice Nexus, I hope you can respect that. Again, so much more I want to say, but little time...

LilyMakesThings commented:
If you want to keep this extension to precisely your (as well as BlueDome's) vision, there's obviously nothing wrong with that. It is your extension (as well as BlueDome's). But so long as you do that, you can't stop others from making an equivalent extension that's more aligned with what _they_ need if you can't adjust it with criticism, and it's for that reason shouldn't be going onto other PRs and claiming them to be redundant by the existence of this extension.

CubesterYT commented:
Ok Window Controls is ready, like actually ready.

NexusKitten commented:
Also, I hate to bring this up again, but are you _sure_ you're giving enough credit to everyone involved in making this extension? I don't know what BlueDome and you discussed, however from the looks of it he made most of the extension, including the move window blocks, resize window blocks, and window title blocks. Actually, in accordance with @LilyMakesThings's and I's advice, I'd consider the original to be superior to this version. However, the only credit I see for BlueDome is right at the bottom of the documentation and feels a little biased (if you read it with a blank slate, it basically says "Mostly by Cubester, and BlueDome did a few things too."
Again, I don't know what you guys have discussed regarding this extension, and if BlueDome has agreed to this, it's fine, however it feels rather rude considering you took the base of their _first_ extension and put your name on it as the main creator.

Azulcat commented:
How to load html?

echoless3484 commented:
> How to load html?

huh

CubesterYT commented:
> How to load html?

Can you elaborate what you want to say please?

Azulcat commented:
> > How to load html?
> 
> Can you elaborate what you want to say please?

So the ext is in html and it does not want to load

CST1229 commented:
> > > How to load html?
> > 
> > 
> > Can you elaborate what you want to say please?
> 
> So the ext is in html and it does not want to load

You're probably trying to load the docs page instead of the actual .js file.

Azulcat commented:
> > > > How to load html?
> > > 
> > > 
> > > Can you elaborate what you want to say please?
> > 
> > So the ext is in html and it does not want to load
> 
> You're probably trying to load the docs page as an extension instead of the actual .js file.

Where can I find the js ext?

CST1229 commented:
> Where can I find the js ext?

https://raw.githubusercontent.com/TurboWarp/extensions/95dd29343e26e705a37d59a7e3528848665b1558/extensions/CubesterYT/WindowControls.js

Azulcat commented:
> > Where can I find the js ext?
> 
> https://raw.githubusercontent.com/CubesterYT/Turbowarp-Extensions/WindowControls/extensions/CubesterYT/WindowControls.js (just note that this will not be loaded unsandboxed as it is not on extensions.turbowarp.org)

I found it thanks anyways for this extension!

CubesterYT commented:
Updated the PR to latest changes of this repo because of the migration to Markdown Docs, but I don't have the time right now to convert them to Markdown, so I updated for now future proof the migration.

CubesterYT commented:
There is one more update I need to do, I'll do that tommorrow.

LilyMakesThings commented:
Why did you move TurboHook so far up?

CubesterYT commented:
Still a couple more things to do, but I gotta wait till I get home.

MilkWithCereal commented:
The wonders of Meta Games begins here.

CubesterYT commented:
Welp, Window Controls V.1.0.0 is officially, truly, done. Even though it already was, the new TW updates allowed me to make it better. Now Window Controls is ready for merging.

GarboMuffin commented:
In some regards I question how necessary some of these blocks are but okay, that''s fine, don't need to change

Last thing is making sure that people are aware that like half of these blocks won't work in the online editor, only desktop app and packaged projects as Electron

CubesterYT commented:
> In some regards I question how necessary some of these blocks are but okay, that''s fine, don't need to change
> 
> Last thing is making sure that people are aware that like half of these blocks won't work in the online editor, only desktop app and packaged projects as Electron

I thought of that, in the documentation at the top, I state the places it mainly works in.

CubesterYT commented:
I'm going to change something later in this updated code, I'll explain why later.

CubesterYT commented:
Then I think I have an idea, I'll have to test it though.

CubesterYT commented:
Ok finished making the changes. Prettier is my new enemy, at least eslint let you keep your own style.

CubesterYT commented:
Wait, let me add an example Project

CubesterYT commented:
Nevermind, I'll do example project later, so it's ready for now.
